snapcraft/_store.py

```diff
@@ -190,6 +189,73 @@ def register(snap_name, is_private=False):
         snap_name))
 
 
+def generate_snap_build(authority_id, snap_id, grade, key_name,
+                        snap_filename):
+    """Return the signed snap-build declaration for a snap on disk."""
+    cmd = [
+        'snap', 'sign-build',
+        '--developer-id=' + authority_id,
+        '--snap-id=' + snap_id,
+        '--grade=' + grade
+    ]
+    if key_name:
+        cmd.extend(['-k', key_name])
+    cmd.append(snap_filename)
+    try:
+        return subprocess.check_output(cmd)
+    except subprocess.CalledProcessError:
+        msg = 'Failed to sign build assertion {}.'.format(assertion)
+        raise snapcraft.internal.meta.CommandError(msg)
+
+
+def sign_build(snap_filename, key_name='default', local=False):
+    if not repo.is_package_installed('snapd'):
+        raise EnvironmentError(
+            'The snapd package is not installed. In order to use '
+            '`sign-build`, you must run `apt install snapd`.')
+
+    if not os.path.exists(snap_filename):
+        raise FileNotFoundError(
+            'The file {!r} does not exist.'.format(snap_filename))
+
+    snap_series = storeapi.constants.DEFAULT_SERIES
+    snap_yaml = _get_data_from_snap_file(snap_filename)
+    snap_name = snap_yaml['name']
+    # test snap has no grade, and so other old snaps that
+    # might need to be asserted once snap-build is supported
+    grade = snap_yaml.get('grade', 'stable')
+
+    store = storeapi.StoreClient()
+    with _requires_login():
+        try:
+            info = store.get_account_information()
+            authority_id = info['account_id']
+            snap_id = info['snaps'][snap_series][snap_name]['snap-id']
+        except KeyError:
+            raise RuntimeError(
+                ('Your account lacks information to sign '
+                 'build assertions for this snap.'))
+
+    snap_build_path = snap_filename + '-build'
+    # XXX cprov 20160922: the local cache/check is needed here
+    # to figure out if the snap-build on disk (named by version)
+    # corresponding to the existing snap blob.
+    if not os.path.isfile(snap_build_path):
```

Comment:
This point has to be better addressed, the current code will push the wrong assertion if the snap was just rebuilt. Without processing the assertion and checking sha3-384, it can only be done by caching the assertion on `~/.config/snapcraft/` as Mark suggested. Even with the local cache, things might get confusing on 3rd-part builds.